### ISSUE TITLE
ITE-5: Consider JWS

### ISSUE THREAD

**Issue by letmaik** [in-toto/ITE#16]
ITE-5 proposes to change the signature envelope to the [SSL Signing Spec](http://github.com/secure-systems-lab/signing-spec). I think this definitely improves on the previous scheme used by in-toto but given that the referenced spec is not an industry standard and no library ecosystem exists for it I wanted to open the discussion on why [JWS](https://tools.ietf.org/html/rfc7515) was rejected.

ITE-5 says "Alternatives such as JWS are extremely complex and error-prone". While there is some truth to it I think it is a bit of an extreme statement, and most issues are not a practical problem anymore in my opinion. I found https://www.scottbrady91.com/JOSE/Alternatives-to-JWTs to be a good objective read on this topic.

ITE-5 also says "Both are also JSON-specific" which I guess is meant to say that it would be preferable if the envelope is not JSON-specific. SSL Signing Spec currently has a single authenticated header field which is the payload type. An optional Key ID  is supported as unauthenticated field. The combination of not using JSON and not base64-encoding the payload and authenticated header for signing allows to use other envelope formats than JSON in their native and most efficient form. In principle, I like this, but in practice it requires that any future additional authenticated or unauthenticated header fields have to be specifically supported by all envelope formats, since there is no automatic mapping. In contrast, JWS defines its header as JSON which makes it extendable without changing the spec, at the expense of being dependent on JSON. It would be nice if JWS would not base64-encode the header and payload before signing, but this is a side issue I think. The central concern I have is the extendable header.

As an aside, not being JSON-specific in the envelope doesn't remove the JSON dependence in the payload. If certain scenarios require a more efficient encoding of signed in-toto statements then I think this has to be considered across all layers. I could image that certain constrained environments would be more happy with a fully binary signed statement rather than a mix of binary and JSON. An example could be a policy engine running on a microcontroller. I could imagine that for this use case a combination of [COSE](https://tools.ietf.org/html/rfc8152) (CBOR equivalent of JWS) and CBOR-encoded in-toto statements (which would have a different media type than it's JSON counterpart) would work. It requires the statement creator to be aware of this but this is an acceptable compromise in those situations I think and can be solved with tooling.

In summary, I want to open up the discussion of using [JWS Compact](https://tools.ietf.org/html/rfc7515#section-3.1) (this is what JWTs are based on) as primary envelope for in-toto statements.

JWS Compact does not support multiple signatures, but I think this is not an issue (more a legacy issue with the older in-toto spec). In practice, claims should be signed by a single party. Even if multiple parties want to sign, then those should be separate documents, since most likely they have slightly different claims. Counter-signing is possible by referencing another in-toto statement as subject in an I-approve-this predicate type. 

The JWS header has the following standard fields to represent the payload type and optional key id:
- `"cty"` which is a media type, e.g. `application/vnd.in-toto+json`
- `"kid"` which is a key id, e.g. `ad91970864102a59765e`

For my own scenarios I would additionally include a non-standard field for the issuer to facilitate key discovery:
- `"issuer"` which can for example be a [Decentralized Identifier](https://www.w3.org/TR/did-core/), e.g. `did:web:letmaik.github.io` (resolving to my [DID doc](https://letmaik.github.io/.well-known/did.json))

**Comment by MarkLodato** [comment]
Personally, I don't have a strong opinion. I originally supported JWS, then I was against it after learning about its complexity, now I'm ambivalent. The arguments made by @letmaik are worth considering.

**Comment by trishankatdatadog** [comment]
Makes a lot of sense, FWIW, considering the existing adoption by the ecosystem. Something like "the good parts of JWS" is worth considering.

**Comment by dlorenc** [comment]
The most compelling argument in favor of JWT's IMO is the existing library/language support.

However, anything we do that's "non-standard" (JWTs, the good parts) would have to be evaluated against the existing libraries to make sure they work.

I'd also argue that the existing library support for JWTs isn't as good as it might seem. I've personally dealt with countless issues with Go JWT libraries for example. Right now there is no recommended, maintained library that has actually been audited. Other languages might be better though.

Building cross-language library support for PAE wouldn't be trivial, but it might not actually be much harder than finding a recommended JWT library that supports custom claims well in every language.

**Comment by MarkLodato** [comment]
Just to clarify, JWT = JWS with a specific payload. The proposal here, as I understand it, is to just use JWS and further restrict to only accept the `"cty"`, `"kid"`, `"alg"`, and maybe `"x5c"` header parameters. (Sadly `"alg"` is required even though it's a bad idea. We'd need to have some guidance around this.)

The question comes down to what is better: a spec that has many buggy implementations, or a spec that has no implementations at all (currently).

I think Dan's suggestion is a good one: do a survey of known JWS implementations and see which ones are maintained, correctly implemented, and would work with our suggested subset of features. Many libraries are JWT-specific, which wouldn't help us.

**Comment by wietse-gmail** [comment]
We may say that 'standard' JWT has many implementations, however 
- The intersection of those implementations with in-toto's needs are small. 
- Some features outside that intersection are unsafe (see https://github.com/google/jws). 
On the other hand, JWT etc. al are a devil that we know. By adopting a new form, we "reset" the history of mistakes to be discovered.

**Comment by MarkLodato** [comment]
To make progress and avoid bikeshedding, I propose that we do not consider this a blocker for ITE-5. We continue to do whatever is necessary to get ITE-5 accepted with signing-spec. During this process, proponents of JWS may convince the community that JWS is enough better to warrant the switch. If they succeed before it is accepted, we update ITE and the implementations. If not, we consider this as "decided." A future ITE can always consider JWS if there is enough interest.

That does raise two questions:
1. What work remains to get ITE-5 accepted?
2. Who would the proponents of JWS have to convince? It is difficult to make a case if you don't know who you are convincing. 😄 

@SantiagoTorres What do you think?

**Comment by kaywilliams** [comment]
The 3T-SBOM/SPDX integrity subgroup (@SantiagoTorres is a co-lead and @adityasaky attends) has proposed an approach to signing SBOMs that uses a subset of the JOSE and COSE specifications. You can find the document [here](https://docs.google.com/document/d/1XZQD82l5Oq_Yr9g1t5vwAZe5UHMCZpHQhJ3L6v6zMVI/edit#). Would it make sense for interested parties to meet and discuss options? That group meets each Monday morning at 8AM Pacific. We can use an upcoming meeting (May 17 is open) for discussion?

@SantiagoTorres, thoughts?

**Comment by joshuagl** [comment]
> The 3T-SBOM/SPDX 3 integrity subgroup (@SantiagoTorres is a co-lead and @adityasaky attends) has proposed an approach to signing SBOMs that uses a subset of the JOSE and COSE specifications. You can find the document [here](https://docs.google.com/document/d/1XZQD82l5Oq_Yr9g1t5vwAZe5UHMCZpHQhJ3L6v6zMVI/edit#). Would it make sense for interested parties to meet and discuss options? That group meets each Monday morning at 8AM Pacific. We can use possibly use an upcoming meeting (May 17 is open).

I'd like to participate in that discussion. Where can I find the details to join the meeting? Are the meetings recorded (May 10th agenda looks interesting, but I have a conflict).

**Comment by kaywilliams** [comment]
> I'd like to participate in that discussion. Where can I find the details to join the meeting? Are the meetings recorded (May 10th agenda looks interesting, but I have a conflict).

Send an email to kayw@microsoft.com and I will forward the recurring meeting invite and have you added to the mailing list. All are welcome.  We haven't recorded meetings so far (but we can). I will record the 5/10 meeting as several people have requested.

**Comment by henkbirkholz** [comment]
Continuing some topic of discussion raised in https://www.youtube.com/watch?v=ct3_U7YFc7k here.

The notion that "if you make some change [to JWS] it is not JWS" is not really accurate, I think. Placement of the header members in either the protected header (is included in signature) or the unprotected header (is not included in signature) is up the user. The only requirement to the format is that at least one type of header must exist. Effectively that header then could even be empty (as all members are always optional in the format).

In JWS, the actual requirements on headers and their memebrs that must be followed are defined by the application - not by the format. You can easily create your own application-specific header members and extend the format. You can also easily increase the authenticity (we mostly say integrity here) by placing them in the protected header. These are all application-specific choices.

All requirements coming from in-toto, its metaformat, and features added by the signing specification here, can be added to the JWS format and kept interoperable with that standard, if the requirements defined are RFC 7515 are satisfied. And to be honest, I think they are pretty easily satisfied.

**Comment by henkbirkholz** [comment]
Picking up a statement from https://github.com/secure-systems-lab/signing-spec/blob/master/README.md

"Why not JOSE/JWS/JWT? JSON-specific, too complicated, too easy to mess up."

JWS is in JSON, but it is definitely not JSON-specific. That seems to be kind of a misleading statement.

Additionally, signing (and especially the meaning of signatures and usage profiles for corresponding key material) is complicated and easy to mess up by the nature of the concept. Tying these "characteristics" to a format is missing the point and honestly reads a little bit like hyperbole, doesn't it?

**Comment by henkbirkholz** [comment]
And a question: while the JWS data model can express the meaning of a signature via its header members, I am not sure how the signing-spec intends to address that. How is that intended to work?

**Comment by henkbirkholz** [comment]
> Sadly `"alg"` is required even though it's a bad idea. We'd need to have some guidance around this.

While that is true and COSE did actually improve in that regard ("This authentication can be done either by placing the [alg] header [member] in the protected header bucket or as part of the externally supplied data"), the alg header member can be placed in the protected header by the application, which would render it temper-evident.

**Comment by MarkLodato** [comment]
> > Sadly `"alg"` is required even though it's a bad idea. We'd need to have some guidance around this.
> 
> While that is true and COSE did actually improve in that regard ("This authentication can be done either by placing the [alg] header [member] in the protected header bucket or as part of the externally supplied data"), in the JWS format the alg header member can be placed in the protected header by the application, which would render it temper-evident.

Putting `alg` in the protected header *does not fix the problem.* It has nothing to do with tamper evidence. See https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/ for more details.

This misunderstanding is further argument against JWS. I'm now more convinced that we should stick with signing-spec. The JWS spec is vulnerable by default; you need to add extra security on top to get something decent. You really are better off starting from scratch.

>  while the JWS data model can express the meaning of a signature via its header members, I am not sure how the signing-spec intends to address that. How is that intended to work?

The `payloadType` field. If you want to add anything else, put it in the `payload`. Most other uses of protected headers are design mistakes. Better to limit the interface to avoid misuse.

> "Why not JOSE/JWS/JWT? JSON-specific, too complicated, too easy to mess up."
>
> JWS is in JSON, but it is definitely not JSON-specific. That seems to be kind of a misleading statement.

You're right, that blurb needs to be updated. JWS allows arbitrary payloads, though it does depend on JSON for the header.

**Comment by MarkLodato** [comment]
@SantiagoTorres we discussed setting a deadline for making a decision. Is that still the plan? Can we pick a date? Originally we discussed June 5 but that might be too soon now, especially considering the upcoming [NIST workshop](https://www.nist.gov/itl/executive-order-improving-nations-cybersecurity/workshop-and-call-position-papers).

I also think it would help to hold a meeting with anyone who has an opinion on this topic.
